deltachat-ios/ViewModel/ChatListViewModel.swift

```diff
@@ -0,0 +1,299 @@
+            filterAndUpdateList(searchText: searchText)
+        } else {
+            // when search input field empty we show default chatList
+            resetSearch()
+        }
+        onChatListUpdate?()
+    }
+
+    private func filterAndUpdateList(searchText: String) {
+
+        // #1 chats with searchPattern in title bar
+        var filteredChatCellViewModels: [ChatCellViewModel] = []
+        var flags: Int32 = 0
+        flags |= DC_GCL_NO_SPECIALS
+        let filteredChatList = dcContext.getChatlist(flags: flags, queryString: searchText, queryId: 0)
+        _ = (0..<filteredChatList.length).map {
+            let chatId = chatList.getChatId(index: $0)
+            let chat = DcChat(id: chatId)
+            let chatName = chat.name
+            let summary = chatList.getSummary(index: $0)
+            let unreadMessages = dcContext.getUnreadMessages(chatId: chatId)
+            let chatTitleIndexes = chatName.containsExact(subSequence: searchText)
+
+            let viewModel = ChatCellViewModel(
+                chatData: ChatCellData(
+                    chatId: chatId,
+                    summary: summary,
+                    unreadMessages: unreadMessages
+                ),
+                titleHighlightIndexes: chatTitleIndexes
+            )
+            filteredChatCellViewModels.append(viewModel)
+        }
+        searchResultsChats.cellData = filteredChatCellViewModels
+
+        // #2 contacts with searchPattern in name or in email
+        var filteredContactCellViewModels: [ContactCellViewModel] = []
+        let contactIds: [Int] = dcContext.getContacts(flags: DC_GCL_ADD_SELF, queryString: searchText)
+
+        let contacts = contactIds.map { return DcContact(id: $0) }
+
+        for contact in contacts {
+            let nameIndexes = contact.displayName.containsExact(subSequence: searchText)
+            let emailIndexes = contact.email.containsExact(subSequence: searchText)
+
+            // contact contains searchText
+            let viewModel = ContactCellViewModel(
+                contactData: ContactCellData(
+                    contactId: contact.id
+                ),
+                titleHighlightIndexes: nameIndexes,
+                subtitleHighlightIndexes: emailIndexes
+            )
+            filteredContactCellViewModels.append(viewModel)
+        }
+        searchResultsContacts.cellData = filteredContactCellViewModels
+
+        // #3 messages with searchPattern (filtered by dc_core)
+        let msgIds = dcContext.searchMessages(searchText: searchText)
+        var filteredMessageCellViewModels: [ChatCellViewModel] = []
+
+
+        for msgId in msgIds {
+            let msg: DcMsg = DcMsg(id: msgId)
+            let chatId: Int = msg.chatId
+            let chat: DcChat = DcChat(id: chatId)
+            let summary: DcLot = msg.summary(chat: chat)
+            let unreadMessages = dcContext.getUnreadMessages(chatId: chatId)
+
+            let viewModel = ChatCellViewModel(
```

Comment:
also here: we create a view model for every search result. this should only happen for the few messages in view.